File: hdf-engine/test/evidence.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { createHash } from 'node:crypto';
import {
  parseEvidencePackage,
  verifyChecksums,
  plannedBaselineRefs,
  coveredBaselineNames,
  coveredBaselinesInPackage,
  agentOverridesInPackage,
  completeness,
  type EvidenceContent,
  type FetchFn,
} from '../src/evidence.js';

// Shared cross-language fixtures (also read by go/evidence_parity_test.go), so
// both evidence-verify cores run the same input.
const evidenceDir = join(dirname(fileURLToPath(import.meta.url)), '..', 'testdata', 'evidence');
const readText = (name: string): string => readFileSync(join(evidenceDir, name), 'utf-8');
const readBytes = (name: string): Uint8Array => readFileSync(join(evidenceDir, name));
const fetch: FetchFn = (uri) => readBytes(uri);
const sha256Hex = (data: Uint8Array): string => createHash('sha256').update(data).digest('hex');

describe('evidence-verify engine — parity with go/evidence.go', () => {
  it('extracts planned baseline refs from the plan (deduped, in order)', () => {
    expect(plannedBaselineRefs(readText('plan.json'))).toEqual(['RHEL9-STIG', 'PostgreSQL-STIG']);
  });

  it('extracts covered baseline names from results', () => {
    expect(coveredBaselineNames(readText('rhel-results.json'))).toEqual(['RHEL9-STIG']);
  });

  it('computes completeness with a sorted missing list', () => {
    const comp = completeness(['RHEL9-STIG', 'PostgreSQL-STIG'], ['RHEL9-STIG']);
    expect(comp.complete).toBe(false);
    expect(comp.missing).toEqual(['PostgreSQL-STIG']);
  });

  it('aggregates covered baselines and agent overrides across the package (same as Go)', () => {
    const { contents } = parseEvidencePackage(readText('package.json'));
    expect(coveredBaselinesInPackage(contents, fetch)).toEqual(['RHEL9-STIG', 'PostgreSQL-STIG', 'RHEL9-STIG']);
    expect(agentOverridesInPackage(contents, fetch)).toBe(2);
  });

  it('classifies checksums: match, mismatch, skipped (same as Go)', () => {
    const goodHash = sha256Hex(readBytes('rhel-results.json'));
    const contents: EvidenceContent[] = [
      { uri: 'rhel-results.json', type: 'hdf-results', checksum: goodHash },
      { uri: 'plan.json', type: 'hdf-plan', checksum: '0'.repeat(64) },
      { uri: 'rhel-results.json', type: 'hdf-results', checksum: '' },
    ];
    const got = verifyChecksums(contents, fetch);
    expect(got.map((r) => r.status)).toEqual(['match', 'mismatch', 'skipped']);
    expect(got[1]?.expected).toBeTruthy();
    expect(got[1]?.actual).toBeTruthy();
  });
});

describe('evidence-verify engine — unit', () => {
  it('parses planRef and contents (checksum value or empty)', () => {
    const pkg = JSON.stringify({
      planRef: 'plan.json',
      contents: [
        { type: 'hdf-plan', uri: 'plan.json', checksum: { algorithm: 'sha256', value: 'aa' } },
        { type: 'hdf-results', uri: 'r.json' },
      ],
    });
    const { planRef, contents } = parseEvidencePackage(pkg);
    expect(planRef).toBe('plan.json');
    expect(contents).toHaveLength(2);
    expect(contents[0]?.checksum).toBe('aa');
    expect(contents[1]?.checksum).toBe('');
  });

  it('classifies a fetch failure as error', () => {
    const throwing: FetchFn = () => {
      throw new Error('no such file: x');
    };
    const got = verifyChecksums([{ uri: 'x', type: 'hdf-results', checksum: 'abc' }], throwing);
    expect(got[0]?.status).toBe('error');
    expect(got[0]?.error).toContain('no such file');
  });

  it('throws on malformed input', () => {
    expect(() => parseEvidencePackage('not json')).toThrow();
    expect(() => plannedBaselineRefs('x')).toThrow();
    expect(() => coveredBaselineNames('x')).toThrow();
  });

  // Go decodes into a typed struct, so a JSON null document leaves zero values
  // and a wrong-typed field is a decode error. The TS peer must not TypeError
  // where Go returns a value or a message.
  it('tolerates a null document, as Go does', () => {
    expect(parseEvidencePackage('null')).toEqual({ planRef: '', contents: [] });
    expect(plannedBaselineRefs('null')).toEqual([]);
    expect(coveredBaselineNames('null')).toEqual([]);
  });

  it('tolerates a null content entry, as Go does', () => {
    const { contents } = parseEvidencePackage(JSON.stringify({ contents: [null, { uri: 'a' }] }));
    expect(contents).toEqual([
      { uri: '', type: '', checksum: '' },
      { uri: 'a', type: '', checksum: '' },
    ]);
    expect(plannedBaselineRefs(JSON.stringify({ assessments: [null, { baselineRef: 'A' }] }))).toEqual(['A']);
    expect(coveredBaselineNames(JSON.stringify({ baselines: [null, { name: 'X' }] }))).toEqual(['X']);
  });

  it('reports a wrong-typed field as an error, as Go does', () => {
    expect(() => parseEvidencePackage('[]')).toThrow(/parse evidence package/);
    expect(() => parseEvidencePackage('"x"')).toThrow(/parse evidence package/);
    expect(() => parseEvidencePackage(JSON.stringify({ contents: 'x' }))).toThrow(/parse evidence package/);
    expect(() => parseEvidencePackage(JSON.stringify({ contents: ['x'] }))).toThrow(/parse evidence package/);
    expect(() => plannedBaselineRefs(JSON.stringify({ assessments: 'x' }))).toThrow(/parse plan/);
    expect(() => coveredBaselineNames(JSON.stringify({ baselines: 'x' }))).toThrow(/parse results/);
  });

  it('reports complete when every planned baseline is covered', () => {
    const comp = completeness(['A'], ['A', 'extra']);
    expect(comp.complete).toBe(true);
    expect(comp.missing).toEqual([]);
  });

  it('tolerates absent optional arrays and fields', () => {
    expect(parseEvidencePackage('{}')).toEqual({ planRef: '', contents: [] });
    // Content present but uri/type absent and checksum object present without a value.
    const { contents } = parseEvidencePackage(JSON.stringify({ contents: [{ checksum: {} }] }));
    expect(contents[0]).toEqual({ uri: '', type: '', checksum: '' });
    expect(plannedBaselineRefs('{}')).toEqual([]);
    expect(coveredBaselineNames('{}')).toEqual([]);
  });

  it('drops assessment/baseline entries missing the field (?? fallback)', () => {
    // An assessment/baseline object present but with baselineRef/name absent must
    // fall back to '' and be filtered out — exercises the nullish-default branch.
    expect(plannedBaselineRefs(JSON.stringify({ assessments: [{ baselineRef: 'A' }, {}] }))).toEqual(['A']);
    expect(coveredBaselineNames(JSON.stringify({ baselines: [{ name: 'X' }, {}] }))).toEqual(['X']);
  });

  it('package aggregators consult only hdf-results entries with a uri and skip unreadable or unparseable ones', () => {
    const agent = new TextEncoder().encode(
      '{"baselines":[{"name":"A","requirements":[{"statusOverrides":[{"appliedBy":{"type":"agent"}}]}]}]}'
    );
    const files: Record<string, Uint8Array> = { 'a.json': agent, 'b.json': new TextEncoder().encode('not json'), 'base.json': agent };
    const mem: FetchFn = (uri) => {
      const data = files[uri];
      if (data === undefined) throw new Error('no such file: ' + uri);
      return data;
    };
    const contents: EvidenceContent[] = [
      { uri: 'a.json', type: 'hdf-results', checksum: '' },
      { uri: 'b.json', type: 'hdf-results', checksum: '' },
      { uri: 'missing.json', type: 'hdf-results', checksum: '' },
      { uri: '', type: 'hdf-results', checksum: '' },
      { uri: 'base.json', type: 'hdf-baseline', checksum: '' },
    ];
    expect(coveredBaselinesInPackage(contents, mem)).toEqual(['A']);
    expect(agentOverridesInPackage(contents, mem)).toBe(1);
    expect(coveredBaselinesInPackage([], mem)).toEqual([]);
  });

  // Go's typed decode rejects a zone-less startTime outright, so this input is
  // the one that pulled the two languages apart. Both must count it.
  it('counts a document whose result timestamps carry no zone', () => {
    const zoneless = JSON.stringify({
      baselines: [
        {
          name: 'A',
          requirements: [
            {
              statusOverrides: [{ appliedBy: { type: 'agent' } }],
              results: [{ status: 'passed', startTime: '2024-01-01T00:00:00' }],
            },
          ],
        },
      ],
    });
    const mem: FetchFn = (uri) => {
      if (uri !== 'z.json') throw new Error('no such file: ' + uri);
      return new TextEncoder().encode(zoneless);
    };
    expect(agentOverridesInPackage([{ uri: 'z.json', type: 'hdf-results', checksum: '' }], mem)).toBe(1);
  });

  it('classifies a non-Error throw as error with the stringified value', () => {
    const throwing: FetchFn = () => {
      throw 'plain string failure';
    };
    const got = verifyChecksums([{ uri: 'x', type: 'hdf-results', checksum: 'abc' }], throwing);
    expect(got[0]?.status).toBe('error');
    expect(got[0]?.error).toBe('plain string failure');
  });
});

File: hdf-engine/go/evidence_test.go
package hdfengine

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"testing"
)

// memFetch builds a FetchFunc over an in-memory uri→bytes map.
func memFetch(files map[string][]byte) FetchFunc {
	return func(uri string) ([]byte, error) {
		b, ok := files[uri]
		if !ok {
			return nil, fmt.Errorf("no such file: %s", uri)
		}
		return b, nil
	}
}

func sha256Hex(b []byte) string {
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:])
}

func TestVerifyChecksums_MismatchAndMatch(t *testing.T) {
	good := []byte(`{"baselines":[{"name":"RHEL9-STIG"}]}`)
	files := map[string][]byte{"good.json": good, "other.json": []byte("other")}
	contents := []EvidenceContent{
		{URI: "good.json", Type: "hdf-results", Checksum: sha256Hex(good)},
		{URI: "other.json", Type: "hdf-results", Checksum: "0000000000000000000000000000000000000000000000000000000000000000"},
		{URI: "nocs.json", Type: "hdf-system", Checksum: ""},
		{URI: "missing.json", Type: "hdf-results", Checksum: "abc"},
	}
	got := VerifyChecksums(contents, memFetch(files))
	if len(got) != 4 {
		t.Fatalf("want 4 results, got %d", len(got))
	}
	if got[0].Status != ChecksumMatch {
		t.Errorf("good.json: status=%s want match", got[0].Status)
	}
	if got[1].Status != ChecksumMismatch {
		t.Errorf("other.json: status=%s want mismatch", got[1].Status)
	}
	if got[1].Expected == "" || got[1].Actual == "" {
		t.Errorf("mismatch must carry expected+actual, got %+v", got[1])
	}
	if got[2].Status != ChecksumSkipped {
		t.Errorf("nocs.json: status=%s want skipped", got[2].Status)
	}
	if got[3].Status != ChecksumError || got[3].Error == "" {
		t.Errorf("missing.json: want error with message, got %+v", got[3])
	}
}

func TestParseEvidencePackage(t *testing.T) {
	pkg := []byte(`{
		"planRef": "plan.json",
		"contents": [
			{"type":"hdf-plan","uri":"plan.json","checksum":{"algorithm":"sha256","value":"aa"}},
			{"type":"hdf-results","uri":"r.json"}
		]
	}`)
	planRef, contents, err := ParseEvidencePackage(pkg)
	if err != nil {
		t.Fatalf("ParseEvidencePackage: %v", err)
	}
	if planRef != "plan.json" {
		t.Errorf("planRef=%q", planRef)
	}
	if len(contents) != 2 {
		t.Fatalf("want 2 contents, got %d", len(contents))
	}
	if contents[0].Checksum != "aa" {
		t.Errorf("contents[0].Checksum=%q want aa", contents[0].Checksum)
	}
	if contents[1].Checksum != "" {
		t.Errorf("contents[1] should have empty checksum, got %q", contents[1].Checksum)
	}
}

func TestParseEvidencePackage_Invalid(t *testing.T) {
	if _, _, err := ParseEvidencePackage([]byte("not json")); err == nil {
		t.Fatal("expected error on malformed package")
	}
}

func TestPlannedBaselineRefs(t *testing.T) {
	plan := []byte(`{"assessments":[{"baselineRef":"A"},{"baselineRef":"B"},{"baselineRef":"A"},{"baselineRef":""}]}`)
	got, err := PlannedBaselineRefs(plan)
	if err != nil {
		t.Fatalf("PlannedBaselineRefs: %v", err)
	}
	if len(got) != 2 || got[0] != "A" || got[1] != "B" {
		t.Fatalf("want [A B] deduped in order, got %v", got)
	}
	if _, err := PlannedBaselineRefs([]byte("x")); err == nil {
		t.Fatal("expected error on malformed plan")
	}
}

func TestCoveredBaselineNames(t *testing.T) {
	results := []byte(`{"baselines":[{"name":"RHEL9-STIG"},{"name":""},{"name":"RHEL9-STIG"}]}`)
	got, err := CoveredBaselineNames(results)
	if err != nil {
		t.Fatalf("CoveredBaselineNames: %v", err)
	}
	if len(got) != 1 || got[0] != "RHEL9-STIG" {
		t.Fatalf("want [RHEL9-STIG] deduped, got %v", got)
	}
	if _, err := CoveredBaselineNames([]byte("x")); err == nil {
		t.Fatal("expected error on malformed results")
	}
}

func TestCompleteness(t *testing.T) {
	comp := Completeness([]string{"A", "B", "C"}, []string{"B"})
	if comp.Complete {
		t.Error("should be incomplete")
	}
	if len(comp.Missing) != 2 || comp.Missing[0] != "A" || comp.Missing[1] != "C" {
		t.Fatalf("missing should be sorted [A C], got %v", comp.Missing)
	}
	full := Completeness([]string{"A"}, []string{"A", "extra"})
	if !full.Complete || len(full.Missing) != 0 {
		t.Fatalf("should be complete, got %+v", full)
	}
}
